### ISSUE TITLE
configuring log4cplus failed

### ISSUE THREAD

**Issue by xiaoshuang-lu** [log4cplus/log4cplus#135]
Version: 1.2.0-RC4
Environment: Visual Studio 2015, debug and release, x64

````{.cpp}
static intmax_t initialize()
{
    ::log4cplus::tistringstream const log4cplusProperties
    (
        ::log4cplus::tstring
        (
            LOG4CPLUS_TEXT
            (
                "log4cplus.rootLogger=TRACE, stdout\n"
                "log4cplus.appender.stdout=log4cplus::ConsoleAppender\n"
                "log4cplus.appender.stdout.Threshold=TRACE\n"
                "log4cplus.appender.stdout.Target=System.out\n"
                "log4cplus.appender.stdout.layout=log4cplus::PatternLayout\n"
                "log4cplus.appender.stdout.layout.ConversionPattern=%D{%Y-%m-%d %H:%M:%S,%q} %p %t %c %L: %m%n\n"
            )
        )
    );

    ::log4cplus::initialize();

    ::log4cplus::PropertyConfigurator
    propertyConfigurator
    (
        * const_cast<::log4cplus::tistringstream *>(& log4cplusProperties)
    );

    // Failed here. Process will be terminated.
    propertyConfigurator.configure();

    return 0;
}

static intmax_t dummy = initialize();

int main(void)
{
    log4cplus::Logger LOGGER = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("int main(void)"));

    LOG4CPLUS_INFO(LOGGER, "who's your daddy?");

    return 0;
}
````

**Comment by wilx** [comment]
First off, you really need to be more precise as to what does not work. I my crystal ball is broken. :)

````{.cpp}
            LOG4CPLUS_TEXT
            (
                "log4cplus.rootLogger=TRACE, stdout\n"
                "log4cplus.appender.stdout=log4cplus::ConsoleAppender\n"
                "log4cplus.appender.stdout.Threshold=TRACE\n"
                "log4cplus.appender.stdout.Target=System.out\n"
                "log4cplus.appender.stdout.layout=log4cplus::PatternLayout\n"
                "log4cplus.appender.stdout.layout.ConversionPattern=%D{%Y-%m-%d %H:%M:%S,%q} %p %t %c %L: %m%n\n"
            )
````

This will not work, IIRC. You have to wrap each individual string literal with `LOG4CPLUS_TEXT()`.

If you are using master branch then you should use the [`log4cplus::Initializer`](https://github.com/log4cplus/log4cplus/blob/master/tests/thread_test/main.cxx#L86) approach instead of calling `log4cplus::initialize()` yourself. If you are using earlier branches, well, you are stuck with this call.

````{.cpp}
static intmax_t dummy = initialize();
````

Do not do this! Initialize at the top of your `main()` function instead (cf. static initialization order fiasco). Also, see my previous comment.

````{.cpp}
log4cplus::Logger LOGGER = log4cplus::Logger::getInstance(LOG4CPLUS_TEXT("int main(void)"));
````

There is no point naming your logger with full signature of your function. Since you are using `PatternLayout` you can use the `%M` specifier to get function name in the output. Better name the logger just `"main"`.

**Comment by wilx** [comment]
Oh, and I have noticed two more issues:

````{.cpp}
::log4cplus::tistringstream const log4cplusProperties
````

Remove the `const`. It is obviously not a constant object.

````{.cpp}
* const_cast<::log4cplus::tistringstream *>(& log4cplusProperties)
````

Do not do this. Just remove the `const` on `log4cplusProperties` at its declaration point.

**Comment by xiaoshuang-lu** [comment]
Hi, thanks for your comments.

1. It would be great if you update the example code posted in http://sourceforge.net/p/log4cplus/wiki/CodeExamples/. ::log4cplus::initialize() was coped from there.

2. The following code snippet has no problems. I have already keep the initialization order in mind.

````{cpp}
static intmax_t dummy = initialize();
````

3. thanks for telling that ````{cpp}%M```` could get the function names.

4. It does not matter whether I specified ````{CPP}const```` here.

By the way, http://sourceforge.net/p/log4cplus/wiki/Home/ hints that both 1.2.0-RC5 and 1.2.0-RC6 are released. But nothing was changed in https://github.com/log4cplus/log4cplus/releases.

Yours sincerely,
Xiaoshuang

**Comment by wilx** [comment]
> It would be great if example code posted in http://sourceforge.net/p/log4cplus/wiki/CodeExamples/ is updated. ::log4cplus::initialize() was coped from there.

Good point. I will do that. TBH, I often forget these examples are there. :)

> By the way, http://sourceforge.net/p/log4cplus/wiki/Home/ hints that both 1.2.0-RC5 and 1.2.0-RC6 are released. But nothing was changed in https://github.com/log4cplus/log4cplus/releases.

I have no idea how this part of the Github web works. Maybe I forgot to push the tags.

**Comment by xiaoshuang-lu** [comment]
Hi Václav Haisman.  1.2.0-RC6 in http://sourceforge.net/p/log4cplus/wiki/Home/ and http://sourceforge.net/p/log4cplus/news/2015/11/log4cplus-120-rc6-released/ shall be 1.1.3-RC6.  Thanks.

**Comment by wilx** [comment]
Oh, and I have noticed two more issues:

````{.cpp}
::log4cplus::tistringstream const log4cplusProperties
````

Remove the `const`. It is obviously not a constant object.

````{.cpp}
* const_cast<::log4cplus::tistringstream *>(& log4cplusProperties)
````

Do not do this. Just remove the `const` on `log4cplusProperties` at its declaration point.

**Comment by wilx** [comment]
> By the way, http://sourceforge.net/p/log4cplus/wiki/Home/ hints that both 1.2.0-RC5 and 1.2.0-RC6 are released. But nothing was changed in https://github.com/log4cplus/log4cplus/releases.

This should be fixed. I always forget to push the tags.

> Hi Václav, 1.2.0-RC6 in http://sourceforge.net/p/log4cplus/wiki/Home/ and http://sourceforge.net/p/log4cplus/news/2015/11/log4cplus-120-rc6-released/ shall be 1.1.3-RC6.  Thanks.

This is fixed now.

**Comment by wilx** [comment]
I am opening #160 to keep track of the code examples update.
